File: java_projects/glue_code/joda-money/src/main/java/org/joda/money/format/MoneyFormatterBuilder.java
package org.joda.money.format;

import org.graalvm.polyglot.Value;
import org.graalvm.polyglot.PolyglotException;
import org.joda.money.ContextInitializer;
import org.joda.money.ExceptionHandler;
import org.joda.money.IntegrationUtils;
import java.util.List;
import java.io.IOException;
import java.util.Locale;
import java.util.ArrayList;
import org.joda.money.BigMoney;
import org.joda.money.CurrencyUnit;
import org.joda.money.IllegalCurrencyException;

public final class MoneyFormatterBuilder {
    private final List<MoneyParser> parsers = new ArrayList<>();
    private final List<MoneyPrinter> printers = new ArrayList<>();
    private static Value clz = ContextInitializer.getPythonClass("/format/MoneyFormatterBuilder.py",
            "MoneyFormatterBuilder");
    private Value obj;

    public MoneyFormatterBuilder(Value obj) {
        this.obj = obj;
    }

    public Value getPythonObject() {
        return obj;
    }

    public MoneyFormatter toFormatter1(Locale locale) {
        //
        // MoneyFormatter.checkNotNull(locale, "Locale must not be null");
        // MoneyPrinter[] printersCopy =
        // (MoneyPrinter[]) printers.toArray(new MoneyPrinter[printers.size()]);
        // MoneyParser[] parsersCopy =
        // (MoneyParser[]) parsers.toArray(new MoneyParser[parsers.size()]);
        // return new MoneyFormatter(0, locale, parsersCopy, printersCopy, null);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("toFormatter1", locale).as(MoneyFormatter.class);
    }

    public MoneyFormatter toFormatter0() {
        //
        // return toFormatter1(Locale.getDefault());
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("toFormatter0").as(MoneyFormatter.class);
    }

    public MoneyFormatterBuilder appendSigned1(
            MoneyFormatter whenPositive, MoneyFormatter whenZero, MoneyFormatter whenNegative) {
        //
        // MoneyFormatter.checkNotNull(whenPositive, "MoneyFormatter whenPositive must
        // not be null");
        // MoneyFormatter.checkNotNull(whenZero, "MoneyFormatter whenZero must not be
        // null");
        // MoneyFormatter.checkNotNull(whenNegative, "MoneyFormatter whenNegative must
        // not be null");
        // SignedPrinterParser pp = new SignedPrinterParser(whenPositive, whenZero,
        // whenNegative);
        // return appendInternal(pp, pp);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendSigned1", whenPositive, whenZero, whenNegative)
                .as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendSigned0(
            MoneyFormatter whenPositiveOrZero, MoneyFormatter whenNegative) {
        //
        // return appendSigned1(whenPositiveOrZero, whenPositiveOrZero, whenNegative);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendSigned0", whenPositiveOrZero, whenNegative).as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder append1(MoneyPrinter printer, MoneyParser parser) {
        //
        // return appendInternal(printer, parser);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("append1", printer, parser).as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder append0(MoneyFormatter formatter) {
        //
        // MoneyFormatter.checkNotNull(formatter, "MoneyFormatter must not be null");
        // formatter.getPrinterParser().appendTo(this);
        // return this;
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("append0", formatter).as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendLiteral(CharSequence literal) {
        //
        // if (literal == null || literal.length() == 0) {
        // return this;
        // }
        // LiteralPrinterParser pp = new LiteralPrinterParser(literal.toString());
        // return appendInternal(pp, pp);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendLiteral", literal).as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendCurrencySymbolLocalized() {
        //
        // return appendInternal(SingletonPrinters.LOCALIZED_SYMBOL, null);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendCurrencySymbolLocalized").as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendCurrencyNumericCode() {
        //
        // return appendInternal(Singletons.NUMERIC_CODE, Singletons.NUMERIC_CODE);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendCurrencyNumericCode").as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendCurrencyNumeric3Code() {
        //
        // return appendInternal(Singletons.NUMERIC_3_CODE, Singletons.NUMERIC_3_CODE);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendCurrencyNumeric3Code").as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendCurrencyCode() {
        //
        // return appendInternal(Singletons.CODE, Singletons.CODE);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendCurrencyCode").as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendAmount1(MoneyAmountStyle style) {
        //
        // MoneyFormatter.checkNotNull(style, "MoneyAmountStyle must not be null");
        // AmountPrinterParser pp = new AmountPrinterParser(style);
        // return appendInternal(pp, pp);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendAmount1", style).as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendAmountLocalized() {
        //
        // AmountPrinterParser pp = new
        // AmountPrinterParser(MoneyAmountStyle.LOCALIZED_GROUPING);
        // return appendInternal(pp, pp);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendAmountLocalized").as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder appendAmount0() {
        //
        // AmountPrinterParser pp =
        // new AmountPrinterParser(MoneyAmountStyle.ASCII_DECIMAL_POINT_GROUP3_COMMA);
        // return appendInternal(pp, pp);
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendAmount0").as(MoneyFormatterBuilder.class);
    }

    public MoneyFormatterBuilder() {
        //

        this.obj = clz.invokeMember("__init__");
    }

    private MoneyFormatterBuilder appendInternal(MoneyPrinter printer, MoneyParser parser) {
        //
        // printers.add(printer);
        // parsers.add(parser);
        // return this;
        //

        // TODO: Check the type mapping below!
        return this.obj.invokeMember("appendInternal", printer, parser).as(MoneyFormatterBuilder.class);
    }

    private static enum Singletons implements MoneyPrinter, MoneyParser {
        NUMERIC_CODE("${numericCode}") {
            @Override
            public void print(MoneyPrintContext context, Appendable appendable, BigMoney money)
                    throws IOException {
                appendable.append(Integer.toString(money.getCurrencyUnit().getNumericCode()));
            }

            @Override
            public void parse(MoneyParseContext context) {
                int count=0;for(;count<3&&context.getIndex()+count<context.getTextLength();count++){char ch=context.getText().charAt(context.getIndex()+count);if(ch<'0'||ch>'9'){break;}}int endPos=context.getIndex()+count;String code=context.getTextSubstring(context.getIndex(),endPos);try{context.setCurrency(CurrencyUnit.ofNumericCode0(code));context.setIndex(endPos);}catch(IllegalCurrencyException ex){context.setError();}
            }
        },
        NUMERIC_3_CODE("${numeric3Code}") {
            @Override
            public void print(MoneyPrintContext context, Appendable appendable, BigMoney money)
                    throws IOException {
                appendable.append(money.getCurrencyUnit().getNumeric3Code());
            }

            @Override
            public void parse(MoneyParseContext context) {
                int endPos=context.getIndex()+3;if(endPos>context.getTextLength()){context.setError();}else{String code=context.getTextSubstring(context.getIndex(),endPos);try{context.setCurrency(CurrencyUnit.ofNumericCode0(code));context.setIndex(endPos);}catch(IllegalCurrencyException ex){context.setError();}}
            }
        },
        CODE("${code}") {
            @Override
            public void print(MoneyPrintContext context, Appendable appendable, BigMoney money)
                    throws IOException {
                appendable.append(money.getCurrencyUnit().getCode());
            }

            @Override
            public void parse(MoneyParseContext context) {
                int endPos=context.getIndex()+3;if(endPos>context.getTextLength()){context.setError();}else{String code=context.getTextSubstring(context.getIndex(),endPos);try{context.setCurrency(CurrencyUnit.of1(code));context.setIndex(endPos);}catch(IllegalCurrencyException ex){context.setError();}}
            }
        };

        public String toString() {
            //
            // return toString;
            //

            // TODO: Check the type mapping below!
            return this.obj.invokeMember("toString").as(String.class);
        }

        private Singletons(String toString) {
            //
            // this.toString = toString;
            //

            this.obj = clz.invokeMember("__init__", toString);
        }

        public void parse(MoneyParseContext context) {
            //
            // int count = 0;
            // for (; count < 3 && context.getIndex() + count < context.getTextLength();
            // count++) {
            // char ch = context.getText().charAt(context.getIndex() + count);
            // if (ch < '0' || ch > '9') {
            // break;
            // }
            // }
            // int endPos = context.getIndex() + count;
            // String code = context.getTextSubstring(context.getIndex(), endPos);
            // try {
            // context.setCurrency(CurrencyUnit.ofNumericCode0(code));
            // context.setIndex(endPos);
            // } catch (IllegalCurrencyException ex) {
            // context.setError();
            // }
            //

            this.obj.invokeMember("parse", context);
        }

        public void print(MoneyPrintContext context, Appendable appendable, BigMoney money)
                throws IOException {
            try {
                //
                // appendable.append(Integer.toString(money.getCurrencyUnit().getNumericCode()));
                //

                this.obj.invokeMember("print", context, appendable, money);
            } catch (PolyglotException e) {
                // TODO: Handle IOException
                throw (IOException) ExceptionHandler.handle(e, "new Singletons(...) { ... }.print");
            }
        }

        private Value obj;

        public Value getPythonObject() {
            return obj;
        }
    }

    private static enum SingletonPrinters implements MoneyPrinter {
    LOCALIZED_SYMBOL;
     private static Value clz = ContextInitializer.getPythonClass("/format/MoneyFormatterBuilder.py", "SingletonPrinters");
    private Value obj;

        public SingletonPrinters(Value obj) {
            this.obj = obj;
        }

        public Value getPythonObject() {
            return obj;
        }

        public String toString() {
            //
            // return "${symbolLocalized}";
            //

            // TODO: Check the type mapping below!
            return this.obj.invokeMember("toString").as(String.class);
        }

        public void print(MoneyPrintContext context, Appendable appendable, BigMoney money)
                throws IOException {
            try {
                //
                // appendable.append(money.getCurrencyUnit().getSymbol1(context.getLocale()));
                //

                this.obj.invokeMember("print", context, appendable, money);
            } catch (PolyglotException e) {
                // TODO: Handle IOException
                throw (IOException) ExceptionHandler.handle(e, "SingletonPrinters.print");
            }
        }
    }
}

File: data/recomposed_projects/deepseek-coder-33b-instruct/body/joda-money/src/main/org/joda/money/format/MoneyFormatterBuilder.py
from __future__ import annotations
import locale
import re
from abc import ABC
from io import IOBase
import io
import typing
from typing import *
import os
from src.main.org.joda.money.format.AmountPrinterParser import *
from src.main.org.joda.money.format.LiteralPrinterParser import *
from src.main.org.joda.money.format.MoneyAmountStyle import *
from src.main.org.joda.money.format.MoneyFormatter import *
from src.main.org.joda.money.format.MoneyParser import *
from src.main.org.joda.money.format.MoneyPrinter import *
from src.main.org.joda.money.format.MultiPrinterParser import *
from src.main.org.joda.money.format.SignedPrinterParser import *
from src.main.org.joda.money.BigMoney import *
from src.main.org.joda.money.CurrencyUnit import *
from src.main.org.joda.money.format.MoneyPrintContext import *
from src.main.org.joda.money.IllegalCurrencyException import *
from src.main.org.joda.money.format.MoneyParseContext import *


class SingletonPrinters(MoneyPrinter):

    LOCALIZED_SYMBOL: SingletonPrinters = None

    def toString(self) -> str:
        return "${symbolLocalized}"

    def print(
        self,
        context: MoneyPrintContext,
        appendable: typing.Union[typing.List, io.TextIOBase],
        money: BigMoney,
    ) -> None:
        appendable.append(money.getCurrencyUnit().getSymbol1(context.getLocale()))


class Singletons(MoneyParser, MoneyPrinter, ABC):

    NUMERIC_CODE: Singletons = None
    NUMERIC_3_CODE: Singletons = None
    CODE: Singletons = None
    __toString: str = ""

    @staticmethod
    def initialize_fields() -> None:
        Singletons.NUMERIC_CODE: Singletons = Singletons("${numericCode}")

        Singletons.NUMERIC_3_CODE: Singletons = Singletons("${numeric3Code}")

        Singletons.CODE: Singletons = Singletons("${code}")

    def toString(self) -> str:
        return self.__toString

    def __init__(self, toString: str) -> None:
        self.__toString = toString

    def __str__(self) -> str:
        return self.__toString

    def print(
        self, context: MoneyPrintContext, appendable: IOBase, money: BigMoney
    ) -> None:
        if self == Singletons.NUMERIC_CODE:
            appendable.write(str(money.getCurrencyUnit().getNumericCode()))
        elif self == Singletons.NUMERIC_3_CODE:
            appendable.write(money.getCurrencyUnit().getNumeric3Code())
        elif self == Singletons.CODE:
            appendable.write(money.getCurrencyUnit().getCode())

    def parse(self, context: MoneyParseContext) -> None:
        if self == Singletons.NUMERIC_CODE:
            count = 0
            for count in range(3):
                if context.getIndex() + count < context.getTextLength():
                    ch = context.getText()[context.getIndex() + count]
                    if ch < "0" or ch > "9":
                        break
                else:
                    break
            endPos = context.getIndex() + count
            code = context.getText()[context.getIndex() : endPos]
            try:
                context.setCurrency(CurrencyUnit.ofNumericCode0(code))
                context.setIndex(endPos)
            except IllegalCurrencyException:
                context.setError()
        elif self == Singletons.NUMERIC_3_CODE:
            endPos = context.getIndex() + 3
            if endPos > context.getTextLength():
                context.setError()
            else:
                code = context.getText()[context.getIndex() : endPos]
                try:
                    context.setCurrency(CurrencyUnit.ofNumericCode0(code))
                    context.setIndex(endPos)
                except IllegalCurrencyException:
                    context.setError()
        elif self == Singletons.CODE:
            endPos = context.getIndex() + 3
            if endPos > context.getTextLength():
                context.setError()
            else:
                code = context.getText()[context.getIndex() : endPos]
                try:
                    context.setCurrency(CurrencyUnit.of1(code))
                    context.setIndex(endPos)
                except IllegalCurrencyException:
                    context.setError()


class MoneyFormatterBuilder:

    __parsers: typing.List[MoneyParser] = []

    __printers: typing.List[MoneyPrinter] = []

    def toFormatter1(self, locale: typing.Any) -> MoneyFormatter:

        MoneyFormatter.checkNotNull(locale, "Locale must not be null")
        printersCopy = [printer for printer in self.__printers]
        parsersCopy = [parser for parser in self.__parsers]
        return MoneyFormatter(0, locale, parsersCopy, printersCopy, None)

    def toFormatter0(self) -> MoneyFormatter:

        locale = Locale.getDefault()
        printersCopy = [printer for printer in self.__printers]
        parsersCopy = [parser for parser in self.__parsers]
        return MoneyFormatter(0, locale, parsersCopy, printersCopy, None)

    def appendSigned1(
        self,
        whenPositive: MoneyFormatter,
        whenZero: MoneyFormatter,
        whenNegative: MoneyFormatter,
    ) -> MoneyFormatterBuilder:

        MoneyFormatter.checkNotNull(
            whenPositive, "MoneyFormatter whenPositive must not be null"
        )
        MoneyFormatter.checkNotNull(
            whenZero, "MoneyFormatter whenZero must not be null"
        )
        MoneyFormatter.checkNotNull(
            whenNegative, "MoneyFormatter whenNegative must not be null"
        )
        pp = SignedPrinterParser(whenPositive, whenZero, whenNegative)
        return self.__appendInternal(pp, pp)

    def appendSigned0(
        self, whenPositiveOrZero: MoneyFormatter, whenNegative: MoneyFormatter
    ) -> MoneyFormatterBuilder:

        MoneyFormatter.checkNotNull(
            whenPositiveOrZero, "MoneyFormatter whenPositiveOrZero must not be null"
        )
        MoneyFormatter.checkNotNull(
            whenNegative, "MoneyFormatter whenNegative must not be null"
        )
        pp = SignedPrinterParser(whenPositiveOrZero, whenPositiveOrZero, whenNegative)
        return self.__appendInternal(pp, pp)

    def append1(
        self, printer: MoneyPrinter, parser: MoneyParser
    ) -> MoneyFormatterBuilder:

        return self.__appendInternal(printer, parser)

    def append0(self, formatter: MoneyFormatter) -> MoneyFormatterBuilder:

        MoneyFormatter.checkNotNull(formatter, "MoneyFormatter must not be null")
        formatter.getPrinterParser().appendTo(self)
        return self

    def appendLiteral(self, literal: str) -> MoneyFormatterBuilder:

        if literal is None or len(literal) == 0:
            return self

        pp = LiteralPrinterParser(literal)
        return self.__appendInternal(pp, pp)

    def appendCurrencySymbolLocalized(self) -> MoneyFormatterBuilder:
        return self.__appendInternal(SingletonPrinters.LOCALIZED_SYMBOL, None)

    def appendCurrencyNumericCode(self) -> MoneyFormatterBuilder:
        return self.__appendInternal(
            SingletonPrinters.NUMERIC_CODE, Singletons.NUMERIC_CODE
        )

    def appendCurrencyNumeric3Code(self) -> MoneyFormatterBuilder:
        return self.__appendInternal(
            SingletonPrinters.NUMERIC_3_CODE, Singletons.NUMERIC_3_CODE
        )

    def appendCurrencyCode(self) -> MoneyFormatterBuilder:
        return self.__appendInternal(SingletonPrinters.CODE, Singletons.CODE)

    def appendAmount1(self, style: MoneyAmountStyle) -> MoneyFormatterBuilder:

        MoneyFormatter.checkNotNull(style, "MoneyAmountStyle must not be null")
        pp = AmountPrinterParser(style)
        return self.__appendInternal(pp, pp)

    def appendAmountLocalized(self) -> MoneyFormatterBuilder:
        pp = AmountPrinterParser(MoneyAmountStyle.LOCALIZED_GROUPING)
        return self.__appendInternal(pp, pp)

    def appendAmount0(self) -> MoneyFormatterBuilder:
        pp = AmountPrinterParser(MoneyAmountStyle.ASCII_DECIMAL_POINT_GROUP3_COMMA)
        return self.__appendInternal(pp, pp)

    def __init__(self) -> None:

        self.__parsers: typing.List[MoneyParser] = []
        self.__printers: typing.List[MoneyPrinter] = []

    def __appendInternal(
        self, printer: MoneyPrinter, parser: MoneyParser
    ) -> MoneyFormatterBuilder:

        self.__printers.append(printer)
        self.__parsers.append(parser)
        return self


Singletons.initialize_fields()
